README
Mobile-P2kMoto

This module uses the p2kmoto library to interface with Motorola Mobile
phones using an USB cable.  It allows access to the filesystem as well
as accessing the SEEM record.

=========================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module is a wrapper around the p2kmoto library
(http://moto4lin.sourceforge.net/). It is known to work under
Linux. The library works under Windows as well, but the module has not
been tested there.

COPYRIGHT AND LICENCE

Copyright (C) 2006-2007 by Mattia Barbon

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


Changes
Revision history for Perl extension Mobile::P2kMoto.

0.03  Sat Jun 28 14:23:36 CEST 2008
	- Fixed the simple filehandle interface to close-on-destroy
	  like normal Perl filehandles.
	- Use Devel::CheckLib to check for p2kmoto library.

0.02  Sun Sep  9 19:40:27 CEST 2007
	- Added easy_openPhone() as a simplified interface
	  to (auto)configuration, P2K mode setting and opening
	  the phone.
	- Added dir() as a simplified and more powerful interface to
	  searchRequest/fileList.
	- Fixed open() and read() to actually work.
	- Added a simplified interface for opening/reading/writing files.
	- Added a simple command line P2K client.

0.01  Tue Jun 26 21:10:14 CEST 2007
	- Initial release


Makefile.PL
#!/usr/bin/perl -w

use strict;
use warnings;
use lib 'inc';

use ExtUtils::MakeMaker;
use Devel::CheckLib;

my $check_lib = !grep /--no-check-lib/, @ARGV;
check_lib_or_exit( lib => 'p2kmoto', header => 'p2kmoto.h' )
    if $check_lib;

eval {
    require ExtUtils::MY_Metafile;
    ExtUtils::MY_Metafile->import;

    my_metafile
      ( { license        => 'gpl',
          dynamic_config => 0,
          } );
};

WriteMakefile
  ( NAME               => 'Mobile::P2kMoto',
    VERSION_FROM       => 'lib/Mobile/P2kMoto.pm',
    PREREQ_PM          => { 'XSLoader' => '0.01',
                            },
    ( $] >= 5.005 ?
      ( ABSTRACT_FROM  => 'lib/Mobile/P2kMoto.pm',
        AUTHOR         => 'Mattia Barbon <mbarbon@cpan.org>' ) : () ),
    LIBS               => [ '-lp2kmoto' ],
    clean              => { FILES => 'const-c.inc const-xs.inc' },
);
if( eval { require ExtUtils::Constant; 1 } ) {
    # If you edit these definitions to change the constants used,
    # you will need to use the generated const-c.inc and const-xs.inc
    # files to replace their "fallback" counterparts before distributing your
    # changes.
    use lib qw(lib);
    use Mobile::P2kMoto::Constants;
    ExtUtils::Constant::WriteConstants
        ( NAME         => 'Mobile::P2kMoto',
          NAMES        => \@Mobile::P2kMoto::Constants::CONSTANTS,
          DEFAULT_TYPE => 'IV',
          C_FILE       => 'const-c.inc',
          XS_FILE      => 'const-xs.inc',
          );
} else {
    use File::Copy;
    use File::Spec;

    foreach my $file ( 'const-c.inc', 'const-xs.inc' ) {
        my $fallback = File::Spec->catfile( 'fallback', $file );
        copy( $fallback, $file ) or die "Can't copy $fallback to $file: $!";
    }
}


example/p2kclient.pl
#!/usr/bin/perl -w

use strict;
use warnings;
use blib;
use Mobile::P2kMoto;
use File::Basename qw();

use constant TIMEOUT => 3000;

print 'Open: ', Mobile::P2kMoto::easy_openPhone( timeout => TIMEOUT ), "\n";

for(;;) {
    print "> ";
    my $cmd = <>;
    chomp $cmd;
    $cmd =~ s/^\s*(\w+)\s*//;
    my $sub = main->can( $1 );

    next unless $sub;
    $sub->( $1, $cmd );
}

sub close {
    my( $cmd, $arg ) = @_;

    print 'Close: ', Mobile::P2kMoto::closePhone(), "\n";
}

sub quit {
    my( $cmd, $arg ) = @_;

    exit 0;
}

sub reboot {
    my( $cmd, $arg ) = @_;

    print 'Reboot: ', Mobile::P2kMoto::reboot(), "\n";
}

sub ls {
    my( $cmd, $arg ) = @_;
    my @files = Mobile::P2kMoto::FS::dir( $arg );

    print $_->size, "\t", $_->attributes, "\t", $_->name, "\n" foreach @files;
}

sub get {
    my( $cmd, $arg ) = @_;
    my( $file ) = Mobile::P2kMoto::FS::dir( $arg );

    unless( $file ) {
        warn "No file found";
        return;
    }

    my $to = File::Basename::basename( $arg );
    my $fh = Mobile::P2kMoto::FS::open( $arg, $file->attributes, $file->size );

    open my $out, '>', $to;
    for(;;) {
        my $buf;
        last if read( $fh, $buf, 10000 ) <= 0;

        print $out $buf;
    }
    CORE::close $out;
    CORE::close $fh;
}

sub put {
    my( $cmd, $arg ) = @_;
    my( $from, $to ) = split ' ', $arg;

    require File::Copy;

    my $out = Mobile::P2kMoto::FS::open( $from, 0, 0 );
    open my $in, '<', $from;

    for(;;) {
        my $buf = '';
        last if read( $in, $buf, 10000 ) == 0;
        print $out $buf;
    }
}

sub rm {
    my( $cmd, $arg ) = @_;

    Mobile::P2kMoto::FS::delete( $arg );
}

sub mkdir {
    my( $cmd, $arg ) = @_;

    Mobile::P2kMoto::FS::createDir( $arg );
}

sub rmdir {
    my( $cmd, $arg ) = @_;

    Mobile::P2kMoto::FS::removeDir( $arg );
}

1;


lib/Mobile/P2kMoto/FS.pm
package Mobile::P2kMoto::FS;

use strict;
use warnings;

sub open {
    my( $file, $mode, $size ) = @_;
    my $rv = _open( $file, $mode );

    require Symbol;
    require Mobile::P2kMoto::FS::FH;

    my $fh = Symbol::gensym();
    tie *$fh, 'Mobile::P2kMoto::FS::FH', $size;

    return $fh;
}

sub dir {
    my( $pattern ) = @_;
    my( $search ) = $pattern;

    my $cb;
    my @files;

    if( $search =~ m{^(?:/[a-zA-Z]/)?[^/]*$} ) {
        $cb = sub { push @files, $_[0] };
    } else {
        $search =~ s{^(/[a-zA-Z]/)?.*/}{$1};
        $pattern =~ s{\*}{.*}g;
        $cb = sub { push @files, $_[0] if $_[0]->name =~ $pattern };
    }

    return () if Mobile::P2kMoto::FS::searchRequest( $search ) < 0;
    return () if Mobile::P2kMoto::FS::fileList( $cb ) < 0;

    return @files;
}

=head1 NAME

Mobile::P2kMoto::FS - filesystem access for Motorola P2K phones

=head1 DESCRIPTION

Functions for listing, creating, reading files and directories.  All
the functions must be called only after a succesful call to
C<openPhone()>.

B<IMPORTANT>: only one file can be open at a given time.

=head1 FUNCTIONS

=head2 open

  my $fh = open( "/a/foo", 0, $size );

Opens or creates a file, returning a Perl filehandle.  Only one file
can be opened at the same time.  Make sure you always call C<close()>
before calling C<open()> again.

Due to the poor interface offered by the C<p2kmoto> library, when
opening a file for reading or updating, the starting size of the file
must be specified.  This might be fixed in the future.

=head2 _open

  my $rv = _open( "/a/foo", 0 );

Opens or creates a file.  Note that since only one file can be opened
at the same time, the return value is a status, not a file handle.
Make sure you always call C<close()> before calling C<open()> again.

This maps directly to the C<p2k_FSAC_open()> function.

Use C<open()> unless you know you need this function.

=head2 close

  my $rv = close();

Closes the currently-open file.

=head2 read

  my $rv = read( $buffer, $size );

Reads C<$size> bytes from the currently-opened file into C<$buffer>.
C<$size> but be less-or-equal than 1024.

=head2 write

  my $rv = write( $buffer, $size );

Write C<$size> bytes from the C<$buffer> info the currently-opened
file.  C<$size> but be less-or-equal than 1024.

=head2 seek

  my $rv = seek( $offset, $whence );

Seeks on the currently-open file.  C<$whence> can be one of
C<P2K_SEEK_BEGIN>, C<P2K_SEEK_CURRENT>, C<P2K_SEEK_END>.  C<$size> but
be less-or-equal than 1024.

=head2 delete

  my $rv = delete( "/a/foo" );

Removes the given file.

=head2 searchRequest

  my $count = searchRequest( "/a/*.mp3" );

Performs a search in the filesystem.  The results can be read using
C<fileList()>.  Note that the search is always recursive.  Returns the
number of files matching the request.

Valid patterns are F<*.foo> and F</c/*.foo>.

=head2 fileList

  my $callback = sub { print $_[0]->name };
  my $rv = fileList( $callback );

Calls C<$callback> for every file in the file list.  If
C<searchRequest()> was not called, lists every file, otherwise only the
ones matched by the last C<searchRequest()>.  The first and only
argument to the callback is a L<Mobile::P2kMoto::FS::FileInfo> object.

=head2 dir

  my @fileinfo = dir( "/a/*.mp3" );

Internally does a call to C<searchRequest> and C<fileList> and returns
a list of L<Mobile::P2kMoto::FS::FileInfo> as result.  Will try to handle
patterns more complicated than those supported by plain C<searchRequest>.

=head2 createDir

  my $rv = createDir( "/a/foodir" );

Creates the directory.

=head2 removeDir

  my $rv = removeDir( "/a/foodir" );

Removes the given directory.  The directory must be empty.

=head2 getVolumeFreeSpace

  my $space = getVolumeFreeSpace( "/a" );

Returns the free space on the volume (in bytes).

=head1 SEE ALSO

L<Mobile::P2kMoto>, L<Mobile::P2kMoto::FS::FileInfo>

=cut

1;


lib/Mobile/P2kMoto/FS/FH.pm
package Mobile::P2kMoto::FS::FH;

use strict;
use warnings;
use base qw(Tie::Handle);

=head1 NAME

Mobile::P2kMoto::FS::FH - internal class used by Mobile::P2kMoto::FS::open

=cut

sub DESTROY {
    my( $self ) = @_;

    CLOSE( $self );
}

sub TIEHANDLE {
    my( $class, $size ) = @_;
    my $self = bless { size => $size, offset => 0 }, $class;

    return $self;
}

sub READ {
    my( $self, undef, $length, $offset ) = @_;
    my( $buf, $res ) = ( '', '' );
    my $remain = $self->{size} - $self->{offset};

    $length = $remain if $remain < $length;

    while( $length ) {
        my $to_read = $length > 1024 ? 1024 : $length;
        my $ok = Mobile::P2kMoto::FS::read( $buf, $to_read );

        if( $ok == 0 ) {
            $res .= $buf;
            $length -= $to_read;
            $self->{offset} += $to_read;
        } else {
            last;
        }
    }

    if( @_ == 4 ) {
        substr( $_[1], $offset ) = $res;
    } else {
        $_[1] = $res;
    }

    return length( $res );
}

#sub READLINE {
#}

sub WRITE {
    my( $self, $buf, $length, $offset ) = @_;

    $buf = substr( $_[1], @_[2 .. 3] ) if @_ > 2;
    $length = length( $buf );

    my $written = 0;
    while( $length ) {
        my $to_write = $length > 1024 ? 1024 : $length;
        my $ok = Mobile::P2kMoto::FS::write( substr( $buf, $written, $to_write ), $to_write );

        if( $ok == 0 ) {
            $written += $to_write;
            $length -= $to_write;
            $self->{offset} += $to_write;
            $self->{size} = $self->{offset} if $self->{offset} > $self->{size};
        } else {
            last;
        }
    }

    return $written;
}

sub CLOSE {
    my( $self ) = @_;

    return unless exists $self->{offset};

    delete $self->{offset};
    delete $self->{size};

    return Mobile::P2kMoto::FS::close() == 0;
}

sub EOF {
    my( $self ) = @_;

    return $self->{offset} == $self->{size};
}

sub SEEK {
    my( $self, $pos, $whence ) = @_;

    if( Mobile::P2kMoto::FS::seek( $pos, $whence ) == 0 ) {
        if( $whence == 0 ) {
            $self->{offset} = $pos;
        } elsif( $whence == 1 ) {
            $self->{offset} += $pos;
        } elsif( $whence == 2 ) {
            $self->{offset} = $self->{size} - $pos;
        }

        return 1;
    } else {
        return 0;
    }
}

sub TELL {
    my( $self ) = @_;

    return $self->{offset};
}

sub FILENO { die "No low-level filehandle" }

1;


lib/Mobile/P2kMoto.pm
package Mobile::P2kMoto;

use 5.006;
use strict;
use warnings;

use Mobile::P2kMoto::FS;
use Mobile::P2kMoto::FS::FileInfo;

our $VERSION = '0.03';
our $AUTOLOAD;

sub AUTOLOAD {
    (my $constname = $AUTOLOAD) =~ s/.*:://;
    my ($error, $val) = constant($constname);
    if ($error)
    {
	require Carp;
	Carp::croak($error);
    }

    no strict 'refs';
    *$AUTOLOAD = sub { $val };
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('Mobile::P2kMoto', $VERSION);

init();

sub easy_openPhone {
    my( %args ) = @_;

    $args{timeout} ||= 3000;
    $args{device}  ||= '/dev/ttyACM0';

    Mobile::P2kMoto::setACMdevice( $args{device} );
    if( $args{at_config} && $args{p2k_config} ) {
        Mobile::P2kMoto::setATconfig( @{$args{at_config}} );
        Mobile::P2kMoto::setP2Kconfig( @{$args{p2k_config}} );
    } else {
        Mobile::P2kMoto::detectPhone();
    }

    my $state = Mobile::P2kMoto::findPhone();

    if( $state == Mobile::P2kMoto::P2K_PHONE_NONE() ) {
        die "No phone found";
    } elsif( $state == Mobile::P2kMoto::P2K_PHONE_AT() ) {
        Mobile::P2kMoto::setP2Kmode( $args{timeout} )
            and die "Failed to set P2K mode";
    } elsif( $state == Mobile::P2kMoto::P2K_PHONE_P2K() ) {
        # do nothing, all is well
    }

    return Mobile::P2kMoto::openPhone( $args{timeout} );
}

1;

__END__

=head1 NAME

Mobile::P2kMoto - interface with Motorola P2K phones

=head1 SYNOPSIS

    use Mobile::P2kMoto;
    use constant TIMEOUT => 3000;

    Mobile::P2kMoto::easy_openPhone( device  => '/dev/ttyACM0',
                                     timeout => TIMEOUT,
                                     );

    Mobile::P2kMoto::FS::searchRequest( '/a/*.wav' );
    Mobile::P2kMoto::FS::fileList( sub { print $_[0]->name, "\n" } );

=head1 DESCRIPTION

This module uses the p2kmoto library to interface with Motorola Mobile
phones using an USB cable.

=head1 FUNCTIONS

=head2 setACMdevice

  setACMdevice( "/dev/ttyACM1" );

The p2kmoto library uses F</dev/ttyACM0> (for Unix) and F<COM3> (for
Win32) as the default device, use this function to change it.

=head2 detectPhone

  my $rv = detectPhone();

Tries to auto-sense a phone connected to the current ACM device.  If
succesful, automatically calls C<setATconfig()> and C<setP2kconfig()>.

=head2 findPhone

  my $rv = findPhone();

Detects current phone state.  Returns one of C<P2K_PHONE_NONE>,
C<P2K_PHONE_AT>, C<P2K_PHONE_P2K>.

=head2 setP2Kmode

  my $rv = setP2Kmode( $timeout );

If the phone is in AT mode, switches it to P2K mode.  Must be called
before accessing the phone via C<openPhone()>.

=head2 openPhone

  my $rv = openPhone( $timeout );

Connects to the phone.  The phone must be in P2K mode before
performing this action.

=head2 easy_openPhone

  my $rv = easy_openPhone( device     => '/dev/ttyACM0',
                           timeout    => 3000,
                           p2k_config => [ 0x1, 0x2 ],
                           acm_config => [ 0x3, 0x4 ],
                           );

A simplified interface to configuring/detecting/opening the phone:
C<device> defaults to C<'/dev/ttyACM0'> and C<timeout> to C<3000>;
C<p2k_config> and C<acm_config> are optional.

This function performa several actions: first it sets the ACM device;
if C<p2k_config> and C<acm_config> are specified, calls
C<setP2Kconfig> and C<setACMconfig>, otherwise calls C<detectPhone>;
switched the phone to P2K mode if it isn't already; calls
C<openPhone>.

=head2 closePhone

  my $rv = closePhone();

Closes the connection to the phone.

=head2 suspend

  my $rv = suspend();

Suspends the phone.

=head2 reboot

  my $rv = reboot();

Reboots the phone.

=head1 CONFIGURATION FUNCTIONS

You should only need the functions below if autodetection fails and if
the defaults values provided by p2kmoto do not work for you.

=head2 setATconfig

  setATconfig( 0x22b8, 0x4902 );

Sets the vendor/product ID used by p2kmoto for the AT interface.

=head2 setP2Kconfig

  setP2Kconfig( 0x22b8, 0x4901 );

Sets the vendor/product ID used by p2kmoto for the P2K interface.

=head2 getACMdevice

  my $device = getACMdevice();

Returns the device used to connect to the phone.

=head2 getATproduct

  my $id = getATproduct();

Sets the product ID used by p2kmoto for the AT interface.

=head2 getATvendor

  my $id = getATvendor();

Sets the vendor ID used by p2kmoto for the AT interface.

=head2 getP2Kproduct

  my $id = getP2Kproduct();

Sets the product ID used by p2kmoto for the P2K interface.

=head2 getP2Kvendor

  my $id = getP2Kvendor();

Sets the vendor ID used by p2kmoto for the P2K interface.

=head1 SEE ALSO

L<Mobile::P2kMoto::FS>, L<Mobile::P2kMoto::FS::FileInfo>

=head1 AUTHOR

Mattia Barbon, E<lt>mbarbon@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006-2007 by Mattia Barbon

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut
